### ISSUE TITLE
Cannot map PEAR merged reads

### ISSUE THREAD

**Issue by jrmerritt6** [FelixKrueger/Bismark#134]
I'm not sure if this is possible, or if there is some modification I can make to Bismark, but I cannot seem to map PEAR merged reads. I have trimmed my data trying Trim Galore!, Trimmomatic, and PEAR, but PEAR gives the best results. I have tried mapping my PEAR assemblies as single end and paired end in Bismark, but mapping efficiency was 0.0%. 

I tried the following code: 
`bismark -n 3 --bowtie1 -e 30 /path/refseq merge.assembled.fastq
bismark -n 3 --bowtie1 --paired -e 30 /path/refseq merge.assembled.fastq`

Part of the reason that I prefer the PEAR assemblies is because I have a substantial amount of fixed genetic divergence in CpGs and expect differential methylation at these SNPs. My reference genome has been n masked at fixed SNPs. 

Is PEAR simply incompatible with Bismark? Or can I modify my code?

**Comment by FelixKrueger** [comment]
Hi @jrmerritt6,

To be honest I have not used PEAR myself before so I don't exactly know what it is doing. I would imagine that overlapping read pairs are merged into a single FastQ entry, but what does it do with reads that do not overlap? Bismark should be able to align merged reads in single-end mode just fine, assuming that PEAR keeps the orientation of the read from Read 1 of a paired-end read (if it would use Read 2 for the overall orientation then you would need `--non_directional` mode). Remaining paired-end reads would still have to be aligned in paired-end mode.

But allow me to ask a stupid question: If you thing that PEAR worked best but you get a 0% mapping efficiency, how badly did the other programs perform? :smile_cat: 

Just quickly regarding your command line:

- For this I would certainly not use `--bowtie1` but use the default (i.e. Bowtie2). 
- Next, `-e 30` would only allow a single mismatch over the entire read which is almost certainly not what you want to do for very long merged reads. 
- Furthermore you should be using only the merged reads in single-end mode
- and align unmerged reads in paired-end mode (with -1 and -2).
- Bowtie1 does not align reads at all when the reference sequence contains Ns

In any case:
1) I would recommend using Trim Galore in the default mode, which should trim just the right thing
```
trim_galore --paired read1.fastq.gz read2.fastq.gz
```

2) I would then align the trimmed data in paired-end mode using a command like this (with the latest versions):
```
bismark --genome /path/refseq -1 read1_trimmed.fq.gz -2 read2_trimmed.fq.gz
```
3) Use/don't use SNPsplit for allele-splitting.

I would be very suprised if this method wouldn't give you very nicely aligned data (with mapping efficiencies even higher than 0.0%!) Let me know how you get on! Cheers, Felix

**Comment by jrmerritt6** [comment]
Felix, 

Thanks so much for you advice! I got alignments much higher than 0.0% 

I ran the PEAR merged assemblies as follows (with a Phred cutoff of 20, which may be a little low): 
`pear -f S84_L001_R1_001.fastq -r S84_L001_R2_001.fastq -o merge -m 600 -q 20
bismark --bowtie2 /path/refseq merge.assembled.fastq`

And then processed the reads using Trim! Galore and Bismark:
```
trim_galore --paired S84_L001_R1_001.fastq S84_L001_R2_001.fastq
bismark --genome /path/refseq/ -1 S84_L001_R1_001_val_1.fq -2 S84_L001_R2_001_val_2.fq
```
In mapping just 4 samples, I found that the PEAR assemblies had higher mapping efficiency than Trim Galore! assemblies: 
PEAR vs. Trim Galore!
56.5% vs. 47.2%
79.0% vs. 65.4%
68.8% vs. 60.4%
72.5% vs. 58.5%

Not sure what accounts for the difference or if there is a way to increase mapping efficiency, but depending on the coverage of my amplicons, this may be sufficient. Thanks again for all of your help!

**Comment by FelixKrueger** [comment]
Two quick questions:

1. The mapping efficiency is relative to all sequences fed in. Is the total number of aligned PEAR sequences higher than the total number of aligned read pairs?

2. To test if there is a potential problem with Read 2 of the pairs, could you try to simply map Read 1 on its own to see what the resulting mapping efficiency is then? The command would be

```
bismark --genome /path/refseq/ S84_L001_R1_001_val_1.fq 
```

Cheers, Felix

**Comment by jrmerritt6** [comment]
1. 3/4 samples had about 11k more reads for PEAR vs. Trim Galore! assemblies. The other sample had 9k more reads in Trim Galore! vs. PEAR assemblies. 
2. Mapping R1s consistently resulted in 9-11k more reads compared to the paired end Trim Galore! assembly. 

These results suggest to me that R2 reads have lower quality than R1 reads, right?

**Comment by FelixKrueger** [comment]
Yes, sounds a bit like that. It almost appears that single-end mapping and PEAR merged reads are roughly equivalent in terms of mapping efficiency, right? Could you attach the FastQC html profiles of a sample R1 and R2?

**Comment by jrmerritt6** [comment]
They are attached. 


[fastqcs.zip](https://github.com/FelixKrueger/Bismark/files/1373204/fastqcs.zip)

**Comment by FelixKrueger** [comment]
The unmerged and merged FastQC reports look pretty similar actually, R2 seems to be slightly worse off than R1. Overall you seem to be dealing with a 2x300bp MiSeq run where the qualities really get quite poor towards the end. Maybe it would be worth increasing the tolerated errors in such long reads a bit, for example by using `--score_min L,0,-0.4` or `--score_min L,0,-0.6`. You seem to be dealing with amplicon data (95% duplication rate), so you even with those relaxed parameters it should be easy to spot whether the alignments go to the intended places.

**Comment by jrmerritt6** [comment]
Thanks for your help! The -0.6 score min aligns ~3,500 more reads than the -0.4 score min. Also, the -0.6 score aligns ~8,000 more reads than the default L,0,-0.2 score min. 

One last question that I'm having a hard time finding online, for amplicon bs-seq, should the reads be deduplicated? If I want to know the proportion methylation and the PCR amplification is related to starting template, shouldn't I actually want the repeated reads? There's not a good reason to think that one sequence would better amplify in the PCR than another.

**Comment by FelixKrueger** [comment]
Hi again. 

Just generally you would assume that all of your reads start and end at the same position, i.e. where the PCR primers bind. Thus you should definitely not de-duplicate or you will lose almost all of your reads. The PCR primers should have been designed so that their binding is not influenced by the methylation status of underlying cytosines (I guess that means they should not cover CG dinucleotides).

Also, depending on your experimental design amplicon libraries sometimes align to the OT and CTOT strands, or OB and CTOB. In that case you would need to set `--non_directional` in order to get all of the alignments. But again, this depends on the setup. You might run a test and if almost all your alignments are against OT then you are already doing the right thing by running Bismark in its default (=directional) mode.

**Comment by jrmerritt6** [comment]
Yep. I'm getting 100% alignment to OT and using the `--non_directional` flag results in exactly the same mapping efficiency. Exactly what I thought re: de-duplication. Thanks for all of your help. I really REALLY appreciate getting advice from an expert on my first NGS project!

**Comment by FelixKrueger** [comment]
You are very welcome! Felix
